File: field/server/src/harness/adapter.js
// The formal Harness Adapter contract.
//
// A harness adapter owns one real external agent process and translates that process's
// native output into Field's canonical event vocabulary (see contracts/field-event-v1.schema.json).
// Both existing adapters — HarnessSession (the direct `claude` CLI, stream-json duplex) and
// KnossosSession (`knossos serve`, NDJSON) — extend this base so the registry, store projection,
// and campaign director can treat any harness uniformly.
//
// Two different "kind" axes exist in this codebase and MUST NOT be conflated:
//   * endpoint/engine kind  — anthropic | openai-compatible | ollama | cameo (endpoints-store.js).
//     Describes the inference backend and its credentials.
//   * harness kind          — the values in HARNESS_KINDS below. Describes HOW Field launches and
//     speaks to the agent process. This is the axis a manifest keys on.
import { EventEmitter, on } from 'node:events';

// The canonical Field event kinds every adapter emits. This is the single source of truth
// the schema-conformance and adapter-conformance tests check against, and it mirrors the
// `kind` enum in contracts/field-event-v1.schema.json.
export const CANONICAL_EVENTS = Object.freeze([
  'session.state',
  'session.message',
  'session.thinking',
  'session.tool_use',
  'session.tool_result',
  'session.usage',
  'session.progress',
  'session.turn_complete',
  'session.verification',
  'session.delegated',
  'session.ended',
  'endpoint.routed',
  'browser.navigated',
  'harness.permission_requested',
]);

// The harness-kind axis. `cli-stream-json` and `ndjson-serve` are the two shipping transports;
// the rest are declared points on the axis reserved for later adapters (do not implement here).
export const HARNESS_KINDS = Object.freeze([
  'cli-stream-json',
  'ndjson-serve',
  'acp',
  'http-api',
  'mcp',
]);

/**
 * The formal adapter interface. Concrete adapters extend this and implement the lifecycle.
 *
 * Contract methods (subclasses MUST implement start/send/cancel; capabilities() is expected):
 *   start(orders, opts?)   -> starts the real process and begins streaming canonical events.
 *   send(input)            -> pushes a user turn / order into the running process.
 *   cancel()               -> terminates the process, emitting `session.ended`.
 *   capabilities()         -> a declarative snapshot (see below) describing what this adapter supports.
 *   stream(opts?)          -> async iterator of { kind, data } canonical events (default provided here).
 *
 * Lifecycle methods the registry also relies on (already implemented by both adapters):
 *   pause() / resume(orders?)  -> stop/restart the process while keeping the session id.
 *   decidePermission(id, decision) -> answer a `harness.permission_requested` (adapters that gate).
 *   beforeStart                -> optional hook the registry assigns for admission/budget checks.
 *
 * Canonical capabilities() shape:
 *   {
 *     kind: <one of HARNESS_KINDS>,
 *     duplex: boolean,        // can accept further turns mid-session
 *     resumable: boolean,     // supports pause()/resume()
 *     permissions: 'mcp-bridge' | 'inline-handshake' | 'none',
 *     delegation: boolean,    // can emit session.delegated
 *     browser: boolean,       // can emit browser.navigated
 *     verification: boolean,  // can emit session.verification
 *     dryRun: boolean,        // supports a read-only / dry-run launch mode
 *   }
 */
export class HarnessAdapter extends EventEmitter {
  /** Canonical emission point: every Field event carries the owning sessionId. */
  emitEvent(kind, data) {
    this.emit('event', kind, { sessionId: this.id, ...data });
  }

  start() {
    throw new Error('HarnessAdapter.start() must be implemented by the adapter');
  }

  send() {
    throw new Error('HarnessAdapter.send() must be implemented by the adapter');
  }

  cancel() {
    throw new Error('HarnessAdapter.cancel() must be implemented by the adapter');
  }

  capabilities() {
    return { kind: null };
  }

  /**
   * Consume this adapter's canonical events as an async iterator. Additive over the
   * existing `'event'` emitter interface — it adds a listener and never changes emission.
   * Pass an AbortSignal to stop iterating.
   */
  async *stream({ signal } = {}) {
    for await (const [kind, data] of on(this, 'event', signal ? { signal } : undefined)) {
      yield { kind, data };
    }
  }
}

/**
 * Validate and normalize a declarative adapter manifest. The manifest is the seam the
 * registry keys on to instantiate an adapter for a given endpoint kind.
 *
 * Manifest shape:
 *   {
 *     id: string,                 // stable adapter id, e.g. 'claude-code'
 *     name: string,               // human label
 *     kind: <one of HARNESS_KINDS>,
 *     endpointKinds: string[],    // endpoint/engine kinds that route to this adapter
 *     launch?: { bin, argv },     // descriptive launch spec (cli-stream-json / ndjson-serve)
 *     endpoint?: object,          // descriptive endpoint spec (http-api / acp / mcp)
 *     auth?: { via, keys },       // how credentials reach the child
 *     capabilities: object,       // matches capabilities() above
 *     Adapter: Function,          // the adapter class/constructor
 *     default?: boolean,          // chosen when no endpointKinds match
 *   }
 */
export function defineAdapterManifest(manifest) {
  const m = manifest ?? {};
  if (!m.id || typeof m.id !== 'string') throw new Error('adapter manifest requires a string id');
  if (!HARNESS_KINDS.includes(m.kind)) {
    throw new Error(`adapter manifest ${m.id} has unknown harness kind: ${m.kind}`);
  }
  if (!Array.isArray(m.endpointKinds)) throw new Error(`adapter manifest ${m.id} requires an endpointKinds array`);
  if (typeof m.Adapter !== 'function') throw new Error(`adapter manifest ${m.id} requires an Adapter constructor`);
  return {
    id: m.id,
    name: m.name ?? m.id,
    kind: m.kind,
    endpointKinds: [...m.endpointKinds],
    launch: m.launch ?? null,
    endpoint: m.endpoint ?? null,
    auth: m.auth ?? null,
    capabilities: m.capabilities ?? {},
    Adapter: m.Adapter,
    default: m.default === true,
  };
}

File: knossos-rs/src/field/adapter.rs
//! The harness adapter contract. Port of `field/server/src/harness/adapter.js`.
//!
//! An adapter owns one real agent process and translates its native output
//! into Field's canonical event vocabulary, so the registry, the projection
//! and the director treat every harness alike. Two "kind" axes exist and
//! must not be conflated: the endpoint kind (anthropic, openai-compatible,
//! cameo) describes the inference backend; the harness kind describes how
//! Field launches and speaks to the agent process.

use serde_json::{json, Value};
use std::path::PathBuf;
use std::sync::Arc;

/// The canonical Field event kinds every adapter emits; mirrors the `kind`
/// enum in `contracts/field-event-v1.schema.json`.
pub const CANONICAL_EVENTS: [&str; 14] = [
    "session.state",
    "session.message",
    "session.thinking",
    "session.tool_use",
    "session.tool_result",
    "session.usage",
    "session.progress",
    "session.turn_complete",
    "session.verification",
    "session.delegated",
    "session.ended",
    "endpoint.routed",
    "browser.navigated",
    "harness.permission_requested",
];

pub const HARNESS_KINDS: [&str; 5] = ["cli-stream-json", "ndjson-serve", "acp", "http-api", "mcp"];

/// Where an adapter's events go: `(kind, data)`, with `sessionId` already
/// stamped into `data`.
pub type EventSink = Arc<dyn Fn(&str, Value) + Send + Sync>;

/// A hook the registry attaches so admission and budget checks run before
/// every (re)start; an error refuses the start.
pub type BeforeStart = Arc<dyn Fn() -> Result<(), String> + Send + Sync>;

/// The mutable identity of a live session, as the registry sees it.
#[derive(Debug, Clone)]
pub struct SessionInfo {
    pub id: String,
    pub agent_id: Option<String>,
    pub name: String,
    pub role: Option<String>,
    pub model: Option<String>,
    pub endpoint_id: Option<String>,
    pub effort: String,
    pub cwd: PathBuf,
    pub workspace_id: Option<String>,
    pub state: String,
}

/// What the registry can do with a session, whichever harness backs it.
pub trait Adapter: Send + Sync {
    fn info(&self) -> SessionInfo;
    fn set_endpoint(&self, endpoint_id: Option<String>, model: Option<String>);
    fn set_effort(&self, effort: &str);
    /// A process (or a process still draining) is owned right now.
    fn is_running(&self) -> bool;
    /// Owned processes, including ones still draining after a stop.
    fn owned_processes(&self) -> usize;
    fn set_before_start(&self, hook: BeforeStart);
    fn start(&self, orders: Option<String>) -> Result<(), String>;
    fn send(&self, text: &str) -> bool;
    fn pause(&self) -> bool;
    fn resume(&self, orders: Option<String>) -> Result<bool, String>;
    fn cancel(&self) -> bool;
    fn decide_permission(&self, request_id: &Value, decision: &str) -> bool;
    fn capabilities(&self) -> Value;
}

/// Stamp `sessionId` into an event payload, the canonical emission point.
pub fn stamp(session_id: &str, data: Value) -> Value {
    let mut out = json!({ "sessionId": session_id });
    if let (Value::Object(o), Value::Object(d)) = (&mut out, data) {
        for (k, v) in d {
            o.insert(k, v);
        }
    }
    out
}
